### PR TITLE
final Task

### PR DESCRIPTION
**Pull request by ViktoriaPa**

### CODE REVIEW

File: allure-results/environment.properties [hunk 1]
```diff
@@ -0,0 +1,4 @@
+Browser=Chrome
```
Comment (shukal94) [inline allure-results/environment.properties]: ignore allure-results folder, this info is specific for your local machine

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/AccountCreationPage.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package com.coherentsolutions.training.auto.web.pashkovskaya.pages;
+
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+public class AccountCreationPage extends BasePage{
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/AccountCreationPage.java:7]: follow encapsulation principle: the fields should be public

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/AccountCreationPage.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    public WebElement firstNameTextField;
+    @FindBy(id = "lastname")
+    public WebElement lastNameTextField;
+    @FindBy(id = "email_address")
+    public WebElement emailTextField;
+    @FindBy(id = "password")
+    public WebElement passwordTextField;
+    @FindBy(id = "password-confirmation")
+    public WebElement passwordConfirmationTextField;
+    @FindBy(xpath = "//button [@class='action submit primary']")
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/AccountCreationPage.java]: 1) this xpath is ambiguous, let's find more obvious locator
2) don't link your xpath to multiple classes, use just one attribute value

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/AccountCreationPage.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    public WebElement createAccountButton;
+
+    public AccountCreationPage(WebDriver driver) {
+        super(driver);
+    }
+
+    public HomePageAuthorizedUser openHomePageAuthorizedUser(String firstName, String lastName, String email, String password, String passwordConf) {
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/AccountCreationPage.java]: let's add more flexibility
1) one action with locator should be in a separate method - single responsibility
2) you never know what else will you have to do with the locator before any action
3) if you want to have signUp method let's just create it as a combination of other methods, example
```
GoogleStartPage
 
Java
public class GoogleStartPage {

  private WebDriver driver;

  @FindBy(id = "searchbar")

  private WebElement searchInput;

  @FindBy(id = "submit")

  private WebElement submitBtn;

  public GoogleStartPage(WebDriver driver) {

    this.driver = driver;

    PageFactory.initElements(this, driver);

  }

  public GoogleSearchResultsPage doSearchQuery(String query) {

    this.setQuery(query);

    return this.submitSearch();

  }

  // implementation
  public boolean isOpened() {

    try {

      return searchInput.isDisplayed();

    } catch (Exception e) {

      return false;

    }

  }

}
```

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/AddNewAddressPage.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    @FindBy(id = "telephone")
+    public WebElement phoneNumberTextField;
+    @FindBy(xpath = "//button[@class = 'action save primary']")
+    public WebElement saveAddressButton;
+
+    public AddNewAddressPage(WebDriver driver) {
+        super(driver);
+    }
+
+    public AddressBookPage saveNewAddress(String address, String city, String zip, String phoneNumber) {
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/AddNewAddressPage.java]: same as for AccountCreationPage

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/AddressBookPage.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+public class AddressBookPage extends BasePage{
+    @FindBy(xpath = "//button[@class='action primary add']")
+    public WebElement addNewAddressButton;
+
+    public AddressBookPage(WebDriver driver) {
+        super(driver);
+    }
+
+    public AddNewAddressPage openAddNewAddressPage() throws InterruptedException {
+        Thread.sleep(10000);
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/AddressBookPage.java]: replace pause with wait

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/AuthorizationPage.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    @FindBy(id = "pass")
+    public WebElement passwordTextField;
+    @FindBy(xpath = "//button [@class='action login primary']")
+    public WebElement signInButton;
+
+    public AuthorizationPage(WebDriver driver) {
+        super(driver);
+    }
+
+    public HomePageAuthorizedUser openHomePageAuthorizedUser(String email, String password) {
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/AuthorizationPage.java]: same comments for AccountCreationPage

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/CartPage.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+import org.openqa.selenium.support.FindBy;
+
+public class CartPage extends BasePage{
+    @FindBy(xpath = "//tr[@class='totals sub'] //span[@class='price']")
+    public WebElement totalPrice;
+    public CartPage(WebDriver driver) {
+        super(driver);
+    }
+
+    public float getTotalPriceFromCart() {
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/CartPage.java]: try to test Assert.assertEquals(1.0f, 0.5f + 0.5f)
never use floating point datatypes with money, java has BigDecimal

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/HomePageAuthorizedUser.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    @FindBy(linkText = "My Account")
+    public WebElement myAccountPage;
+    @FindBy(linkText = "Women")
+    public WebElement womenLink;
+
+    public HomePageAuthorizedUser(WebDriver driver) {
+        super(driver);
+    }
+
+    public MyAccountPage openMyAccountPage() throws InterruptedException {
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/HomePageAuthorizedUser.java]: replace pause with wait and same comments for AccountCreationPage

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/HomePageAuthorizedUser.java [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+        Thread.sleep(2500);
+        loggedInUserLabel.click();
+        myAccountPage.click();
+
+        return new MyAccountPage(driver);
+    }
+
+    public WomenClothesPage openWomenClothesPage() {
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/HomePageAuthorizedUser.java]: is there any difference between WomenClothesPage and other pages except content?

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/HomePageAuthorizedUser.java [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+public class HomePageAuthorizedUser extends BasePage{
+    @FindBy(xpath = "//div[@class = 'panel header'] //button[@class = 'action switch']")
+    public WebElement loggedInUserLabel;
+    @FindBy(linkText = "My Account")
+    public WebElement myAccountPage;
+    @FindBy(linkText = "Women")
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/HomePageAuthorizedUser.java]: let's unify this locator for another options

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/ItemPage.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    public void addProductToCart() {
+        sizeLabel.click();
+        colorLabel.click();
+        addToCartLink.click();
+    }
+
+    public WishListPage openWishListPage() {
+        addToWishListLink.click();
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/ItemPage.java]: does it redirect to wishlist immediately?

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/ItemPage.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    @FindBy(id = "product-addtocart-button")
+    public WebElement addToCartLink;
+    @FindBy(xpath = "//*[@id='maincontent']/div[2]/div/div[1]/div[5]/div/a[1]")
+    public WebElement addToWishListLink;
+
+    public ItemPage(WebDriver driver) {
+        super(driver);
+    }
+
+    public void addProductToCart() {
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/ItemPage.java]: same comments for AccountCreationPage

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/WomenClothesPage.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    @FindBy(xpath = "//a[@class='action showcart']")
+    public WebElement cartIcon;
+    @FindBy(xpath = "//a[@class='action viewcart']")
+    public WebElement cartLink;
+
+    public WomenClothesPage(WebDriver driver) {
+        super(driver);
+    }
+
+    public float addThreeProductsToCart() throws InterruptedException {
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/WomenClothesPage.java]: let's change API: I want method that will choose product by its name

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/WomenClothesPage.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
+package com.coherentsolutions.training.auto.web.pashkovskaya.pages;
+
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+public class WomenClothesPage extends BasePage{
+    @FindBy(xpath = "//ol[@class='product-items widget-product-grid'][1]")
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/WomenClothesPage.java]: only one product?

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/WomenClothesPage.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    public WebElement productItem;
+    @FindBy(xpath = "//a[@class='action showcart']")
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/WomenClothesPage.java]: same for xpath comment

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/util/TestAllureListener.java [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Override
+    public void onTestSuccess(ITestResult iTestResult) {
+
+    }
+
+    @Override
+    public void onTestFailure(ITestResult iTestResult) {
+        System.out.println("I am in onTestFailure method " + getTestMethodName(iTestResult) + " failed");
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/util/TestAllureListener.java]: add logger

File: src/test/java/com/coherentsolutions/training/auto/web/pashkovskaya/tests/TestAccountCreation.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+
+import static com.coherentsolutions.training.auto.web.pashkovskaya.util.MainConstants.*;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+@Listeners({ TestAllureListener.class })
+public class TestAccountCreation extends BaseTest {
+    public static final By PASSWORD_MISMATCH_ERROR = By.id("password-confirmation-error");
+    public static final String PASSWORD_MISMATCH_MESSAGE = "Please enter the same value again.";
+    public static final String WRONG_PASSWORD = "VikaVqwe1234";
+    public static final By ACCOUNT_HAS_BEEN_ALREADY_CREATED_ERROR_MESSAGE = By.xpath("//div[@class='messages']");
```
Comment (shukal94) [inline src/test/java/com/coherentsolutions/training/auto/web/pashkovskaya/tests/TestAccountCreation.java]: why do you need By if you have page objects?

File: src/test/java/com/coherentsolutions/training/auto/web/pashkovskaya/tests/TestAccountCreation.java [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+    @Test(groups = {"AccountCreation"}, description = "Creation a new account. Entered passwords are different")
+    public void testPasswordsAreDifferent() {
+        WebDriver driver = PageDriver.getDriver();
+        driver.get(LINK_TO_MAIN_PAGE);
+
+        HomePageUnauthorizedUser homePageUnauthorizedUser = new HomePageUnauthorizedUser(driver);
+        AccountCreationPage accountCreationPage = homePageUnauthorizedUser.openAccountCreationPage();
+        accountCreationPage.openHomePageAuthorizedUser(FIRST_NAME, LAST_NAME, EMAIL, PASSWORD, WRONG_PASSWORD);
+
+        assertEquals(driver.findElement(PASSWORD_MISMATCH_ERROR).getText(), PASSWORD_MISMATCH_MESSAGE, "Error message mismatch");
```
Comment (shukal94) [inline src/test/java/com/coherentsolutions/training/auto/web/pashkovskaya/tests/TestAccountCreation.java]: move it into page

**Review by shukal94 — CHANGES_REQUESTED** [review]

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/ItemPage.java (hunk 12 above)
Comment (ViktoriaPa) [reply src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/ItemPage.java]: yes

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/HomePageAuthorizedUser.java (hunk 10 above)
Comment (ViktoriaPa) [reply src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/HomePageAuthorizedUser.java]: As far as I understand, there is no difference only between the WomenClothesPage and MenClothesPage. I don't use MenClothesPage, so I didn't add it in my project

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/WomenClothesPage.java (hunk 15 above)
Comment (ViktoriaPa) [reply src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/WomenClothesPage.java]: I need one of any product to test adding to the wish list or to the cart. This locator is the locator of the first available product in the list. 

As I understood, it doesn't matter which product I choose, so I chose the first one

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/AddNewAddressPage.java [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+        enterAddress(address);
+        enterCity(city);
+        openCountryDropdown();
+        selectCountry();
+        openStateDropdown();
+        selectState();
+        enterZip(zip);
+        enterPhoneNumber(phoneNumber);
+        clickSaveAddressButton();
+        driver.manage().timeouts().implicitlyWait(5, TimeUnit.SECONDS);
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/AddNewAddressPage.java]: statement will not have effect here

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/AuthorizationPage.java [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    public void clickSignInButton(){
+        signInButton.click();
+    }
+
+    public HomePageAuthorizedUser openHomePageAuthorizedUser(String email, String password) {
+        driver.manage().timeouts().implicitlyWait(10, TimeUnit.SECONDS);
+        enterEmail(email);
+        driver.manage().timeouts().implicitlyWait(3, TimeUnit.SECONDS);
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/AuthorizationPage.java]: see above

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/CartPage.java (hunk 8 above)
Comment (shukal94) [reply src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/CartPage.java]: java has BigDecimal type for money

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/HomePageAuthorizedUser.java [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+    public void clickCartIcon(){
+        WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(30));
+        wait.until(ExpectedConditions.visibilityOf(counterOfAddedToCartItems));
+
+        cartIcon.click();
+    }
+    public void clickCartLink(){
+        cartLink.click();
+    }
+    public float addThreeProductsToCart() {
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/HomePageAuthorizedUser.java]: let's validate products one by one. one method call - one product to check

File: src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/ItemPage.java [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+import org.openqa.selenium.support.FindBy;
+
+public class ItemPage extends BasePage{
+    @FindBy(xpath = "//div[@class='swatch-attribute size'] /div[@class='swatch-attribute-options clearfix'] /div[1]")
+    private WebElement sizeLabel;
+    @FindBy(xpath = "//div[@class='swatch-option color'][1]")
+    private WebElement colorLabel;
+    @FindBy(id = "product-addtocart-button")
+    private WebElement addToCartLink;
+    @FindBy(xpath = "//*[@id='maincontent']/div[3]/div/div[2]/div[3]/div/div/ol/li[1]/div/div/div[4]/div/div[2]/a[1]")
```
Comment (shukal94) [inline src/main/java/com/coherentsolutions/training/auto/web/pashkovskaya/pages/ItemPage.java]: do not use autogenerated xpathes

**Review by shukal94 — CHANGES_REQUESTED** [review]

**Review by shukal94 — APPROVED** [review]

**Pull request merged**